src/engine-actions/translate-swap.js

```diff
@@ -38,40 +39,56 @@ async function translateSwap (takerAddress, swapHash, amount, maxTime, finalCltv
   this.logger.info(`Translating swap for ${swapHash} by sending to ${takerAddress}`,
     { amount, maxTime })
 
-  const { publicKey } = networkAddressFormatter.parse(takerAddress)
+  // To ensure translateSwap is idempotent, we first make a call to getPaymentPreimage to make sure we don't have an
+  // existing payment that is either In-Flight, Completed, or Failed. If there is not an existing preimage for the
+  // provided swapHash, we can be confident that a call to sendPayment will result in creating only one downstream
+  // payment for this swap
+  try {
+    const { paymentPreimage, permanentError } = await getPaymentPreimage(swapHash)
```

Comment:
might be worth having this return the preimage and `throw` to match the style defined in this file.